backend/de.metas.swat/de.metas.swat.base/src/main/java/org/adempiere/inout/util/ShipmentScheduleQtyOnHandStorage.java

```diff
@@ -33,29 +14,45 @@
 import de.metas.material.cockpit.stock.StockDataQuery;
 import de.metas.material.cockpit.stock.StockRepository;
 import de.metas.material.event.commons.AttributesKey;
-import org.eevolution.api.PPOrderId;
-import org.eevolution.api.QtyCalculationsBOM;
-import org.eevolution.api.QtyCalculationsBOMLine;
 import de.metas.product.ProductId;
 import de.metas.util.Services;
 import lombok.NonNull;
 import lombok.ToString;
+import org.adempiere.util.lang.impl.TableRecordReference;
+import org.adempiere.warehouse.WarehouseId;
+import org.adempiere.warehouse.api.IWarehouseDAO;
+import org.compiere.util.Util.ArrayKey;
+import org.eevolution.api.IPPOrderBL;
+import org.eevolution.api.PPOrderId;
+import org.eevolution.api.QtyCalculationsBOM;
+import org.eevolution.api.QtyCalculationsBOMLine;
+import org.slf4j.MDC.MDCCloseable;
+
+import javax.annotation.Nullable;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Stream;
 
 /**
  * Loads stock details which are relevant to given {@link I_M_ShipmentSchedule}s.
  * Allows to change (in memory!) the qtyOnHand.
  */
 @ToString(of = "stockDetails")
-public class ShipmentScheduleQtyOnHandStorage
+public class ShipmentScheduleQtyOnHandStorage implements IShipmentScheduleQtyOnHandStorage
 {
 	// services
 	private final IShipmentScheduleEffectiveBL shipmentScheduleEffectiveBL = Services.get(IShipmentScheduleEffectiveBL.class);
 	private final IWarehouseDAO warehouseDAO = Services.get(IWarehouseDAO.class);
 	private final IPPOrderBL ppOrdersBL = Services.get(IPPOrderBL.class);
 
 	private final ImmutableList<ShipmentScheduleAvailableStockDetail> stockDetails;
-	private final Map<ArrayKey, StockDataQuery> cachedMaterialQueries = new HashMap<>();
-	private final Map<PPOrderId, Optional<QtyCalculationsBOM>> cachedPickingBOMs = new HashMap<>();
+	private final CCache<ArrayKey, StockDataQuery> cachedMaterialQueries = CCache.newLRUCache("QtyOnHandCachedMaterialQueries", 200, CCache.EXPIREMINUTES_Never);
```

Comment:
why using CCache on an object with very short live?
i.e. ShipmentScheduleQtyOnHandStorage is instantiated by ShipmentScheduleQtyOnHandProvider.getStorageFor on demand.

Reply:
changed back to Map.